logstash-core/lib/logstash/java_pipeline.rb

```diff
@@ -376,21 +376,23 @@ def inputworker(plugin)
   # this method is intended to be called from outside the pipeline thread
   # @param before_stop [Proc] code block called before performing stop operation on input plugins
   def shutdown(&before_stop)
+    return if finished_execution?
     # shutdown can only start once the pipeline has completed its startup.
     # avoid potential race condition between the startup sequence and this
     # shutdown method which can be called from another thread at any time
     sleep(0.1) while !ready?
 
     # TODO: should we also check against calling shutdown multiple times concurrently?
 
-    before_stop.call if block_given?
+    background_thread = Thread.new { before_stop.call } if block_given?
 
     stop_inputs
 
     # We make this call blocking, so we know for sure when the method return the shutdown is
     # stopped
     wait_for_workers
     clear_pipeline_metrics
+    background_thread.exit if background_thread
```

Comment:
Should we at least make a time-limited attempt to join the background thread before sending `Thread#exit`?

Reply:
good point, this should actually just be a .wait, and the shutdownwatcher should look for `pipeline.finished_execution?` as exit condition.

Reply:
I modified this to use thread.join while also modifying the ShutdowWatcher loop to check if `pipeline.finished_execution?` is true.